flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/sink/FullChangelogStoreWriteOperator.java

```diff
@@ -0,0 +1,224 @@
+        super.initializeState(context);
+
+        TupleSerializer<Tuple3<Long, BinaryRowData, Integer>> writtenBucketStateSerializer =
+                new TupleSerializer<>(
+                        (Class<Tuple3<Long, BinaryRowData, Integer>>) (Class<?>) Tuple3.class,
+                        new TypeSerializer[] {
+                            LongSerializer.INSTANCE,
+                            new BinaryRowDataSerializer(
+                                    table.schema().logicalPartitionType().getFieldCount()),
+                            IntSerializer.INSTANCE
+                        });
+        writtenBucketState =
+                context.getOperatorStateStore()
+                        .getListState(
+                                new ListStateDescriptor<>(
+                                        "table_store_written_buckets",
+                                        writtenBucketStateSerializer));
+        writtenBuckets = new TreeMap<>();
+        writtenBucketState
+                .get()
+                .forEach(
+                        t ->
+                                writtenBuckets
+                                        .computeIfAbsent(t.f0, k -> new HashSet<>())
+                                        .add(Tuple2.of(t.f1, t.f2)));
+
+        snapshotIdentifierToCheck = null;
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        currentWrittenBuckets = new HashSet<>();
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        SinkRecord record = writeRecord(element);
+
+        BinaryRowData partition = record.partition();
+        int bucket = record.bucket();
+        // partition is a reused BinaryRowData
+        // we first check if the tuple exists to minimize copying
+        if (!currentWrittenBuckets.contains(Tuple2.of(partition, bucket))) {
+            currentWrittenBuckets.add(Tuple2.of(partition.copy(), bucket));
+        }
+    }
+
+    @Override
+    public void snapshotState(StateSnapshotContext context) throws Exception {
+        super.snapshotState(context);
+
+        List<Tuple3<Long, BinaryRowData, Integer>> writtenBucketList = new ArrayList<>();
+        for (Map.Entry<Long, Set<Tuple2<BinaryRowData, Integer>>> entry :
+                writtenBuckets.entrySet()) {
+            for (Tuple2<BinaryRowData, Integer> bucket : entry.getValue()) {
+                writtenBucketList.add(Tuple3.of(entry.getKey(), bucket.f0, bucket.f1));
+            }
+        }
+        writtenBucketState.update(writtenBucketList);
+    }
+
+    @Override
+    protected List<Committable> prepareCommit(boolean blocking, long checkpointId)
+            throws IOException {
+        if (snapshotIdentifierToCheck != null) {
+            Optional<Snapshot> snapshot = findSnapshot(snapshotIdentifierToCheck);
+            if (snapshot.map(s -> s.commitKind() == Snapshot.CommitKind.COMPACT).orElse(false)) {
```

Comment:
It looks like `AbstractFileStoreWrite` is dealing with the same problem, one is to clear `writer`, one is to clear bucket information.

Can we have a common class to deal with this? Otherwise, this code is a little hard to understand.

Reply:
They are actually very different.

`AbstractFileStoreWrite` needs to determine if all written records have been committed, no matter what the snapshot kind is. So we only need to compare with the latest commit identifier by this user.

However `FullChangelogStoreWriteOperator` must find the exact snapshot it is looking for. It is possible that the latest commit identifier is newer, but no compact changes is committed due to conflicts.